### ISSUE TITLE
Problem: the "normalize for thumbnail" micro-service has inefficiencies and redundancies

### ISSUE THREAD

**Issue by jrwdunham** [artefactual/archivematica#1022]
Proposed solution: reduce and/or remove some "normalize for thumbnail" tasks. Some normalization can be avoided altogether, e.g., the creation of thumbnails for XML files. See also:

- https://github.com/artefactual-labs/archivematica-acceptance-tests/issues/62
- https://docs.google.com/document/d/1Fv8SthGKndeIg_TzVNNXXJUg89Rm0PYC80ebf50gIuI/edit

**Comment by jhsimpson** [comment]
Some additional detail:

The Normalization Microservice includes a job that creates thumbnail images for every file. These thumbnails are used in the Archival Storage tab when the 'show files' checkbox is clicked.  

If search indexing is disabled (a new feature in v1.7.0) then the thumbnails can't be accessed, but they are still created.

Thumbnails often aren't unique. A thumbnail for an xml file just tells you it is an xml file. The current implementation physically copies a placeholder image once for every file.  The same thumbnail image could be use for every file.

This job within the Microservice should be configurable so that users can determine not to run it. This will save processing time and reduce the size of preserved datasets.

Some analysis is required to determine best model for configuring this (i.e. as part of FPR rule, or as part of processing configuration?)

Acceptance Criteria
produce a decision matrix showing options for how to make this microservice configurable
implement a first iteration of the selection solution
test to demonstrate impact of the change (performance and preservation quality).

This issue has been filed in the JiscRDSS fork of Archivematica as well: https://github.com/JiscRDSS/archivematica/issues/40

**Comment by jambun** [comment]
We looked at this today and walked through how we might implement it using the processing configuration to set pass a purpose to the normalize script to tell it not to generate thumbnails. It seems like the right approach, but maybe we're not seeing and alternative. It seems to us that using an FP rule would not help in this case. But again, we might be missing a subtlety here.

Should we go ahead and implement this?

**Comment by jambun** [comment]
Thanks @sromkey - that sounds like a good plan.

@jhsimpson - shall we go ahead with the first part of Sarah's suggested approach?

**Comment by jhsimpson** [comment]
I like the idea of doing a couple or three iterations.  How does this sound?

Iteration 1:
add a new question to the processing configuration page called 'generate thumbnails' with three options (Yes, No, None).  None allows the user to choose during processing.

Iteration 2:
add another option to the 'generate thumbnails' question ('Yes, skip default rule') which does generate thumbnails, but only when there is a valid FPR command that can generate a meaningful thumbnail.

Iteration 3: 
remove the current 'Normalize' question (which has 6 possible options) and replace it with 2 questions:
 - 'generate preservation derivatives' with 4 options ('Yes', 'Use user provided files', 'No', 'None')
(Use user provided files == normalize manually, yes means use FPR rules
- 'generate access derivatives with 4 options ('Yes, using original files', 'Yes, using service files', 'No', 'None')

There are a few goals that overlap, that this work addresses partially or in full, including but not limited to:
- make all preservation actions individually configurable  (this is a goal that iteration 3 would meet) 
- save the processing time wasted on thumbnail generation on servers configured with search indexing disabled (iteration 1 addresses this)
- save the wasted disk space and io caused by copying thousands of copies of the default thumbnail when no meaningful thumbnails can be generated (iteration 2 addresses this).

I think iteration 3 requires some more requirements analysis, it would require a significant change to the workflow I think, so starting with iteration 1 is a good idea.

**Comment by jambun** [comment]
Thanks @jhsimpson - we'll start on iteration 1

**Comment by joel-simpson** [comment]
parent to #33

**Comment by joel-simpson** [comment]
parent to https://github.com/archivematica/Issues/issues/34

**Comment by joel-simpson** [comment]
Testing the solution in qa/1.x built on 24 July. To test this feature, we need to test the 3 different modes ("yes", "no" and "yes, without defaults")

1) Selecting "Yes" to the Choose thumbnail mode decision point should generat thumbnails for all original objects in the Transfer. I tested with the SampleTransfers/Images directory and confirmed that Generate Thumbnails microservice ran successfully, all tasks were documented with stdout as expected, and I could see 10 thumbnails (corresponding to 10 original objects) after downloading the AIP. 
I also used SampleTransfers/BagTransfer to create both an AIP and a DIP. thumbnails were generated and confirmed I could see them in both the AIP and the DIP (with 6 thumbnails corresponding to the 6 original files).
 
2) Selecting "No" to this decision point should not generate any Thumbnails. I did this with SampleTransfers/Images; once when only normalizing for Preservation, and once normalising for Preservation and Access. In both cases all microservices succeeded as expected and there were no Thumbnails in the AIPs or DIP produced. 

3) Selecting "Yes, without defaults" should produce thumbnails for every file that has a corresponding FPR Rule. In SampleTransfers/Images, there are 10 images in total each with a different format. There are 7 FPR rules for Normalization, where Purpose = Thumbnails that relate to the formats in that Transfer. After testing I was able to confirm that there were 7 thumbnails produced as expected. 

It is worth noting that in doing all of these tests, I did some by selecting the decision point manually (where the process config setting is "none") and some where I set the decision point directly in the processing configuration.

In short, looks good to me! marking this as verified.
